# Keras Tuner
[Keras Tuner](https://keras.io/api/keras_tuner/) is an open-source python library developed exclusively for tuning the hyperparameters of ANN and CNN. Keras tuner currently supports four types of tuners or algorithms namely,
1. Bayesian Optimization
2. Hyperband
3. Sklearn
4. Random Search

You can instakk the Keras tyner on your system using the following command,

In [ ]:
#!pip install keras-tuner

In this tutorial we will do regression on boston housing dataset. Firstly, without using Keras Tuner and then using keras tuner.
## Without Keras Tuner

In [1]:
import tensorflow as tf
import keras_tuner
from tensorflow import keras

In [2]:
from tensorflow.keras.datasets import boston_housing
(X_train, y_train), (X_test, y_test) = boston_housing.load_data()

65536/57026 [==================================] - 0s 0us/step


In [3]:
from sklearn.preprocessing import StandardScaler
from tensorflow.keras import models, layers

In [4]:
# set random seed
from numpy.random import seed
seed(42)
import tensorflow
tensorflow.random.set_seed(42)

# preprocessing - normalization
scaler = StandardScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [6]:
# model building
model = models.Sequential()
model.add(layers.Dense(8, activation='relu', input_shape=(X_train.shape[1],)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dropout(0.1))
model.add(layers.Dense(1))

In [7]:
# compile model using rmsprop
model.compile(optimizer='rmsprop',loss='mse',metrics=['mse'])

In [8]:
# model training
history = model.fit(X_train_scaled, y_train, validation_split=0.2, epochs=10)

Epoch 1/10
11/11 [==============================] - 3s 71ms/step - loss: 572.4401 - mse: 572.4401 - val_loss: 636.1214 - val_mse: 636.1214
Epoch 2/10
11/11 [==============================] - 0s 15ms/step - loss: 559.4045 - mse: 559.4045 - val_loss: 623.9985 - val_mse: 623.9985
Epoch 3/10
11/11 [==============================] - 0s 14ms/step - loss: 548.9240 - mse: 548.9240 - val_loss: 613.4424 - val_mse: 613.4424
Epoch 4/10
11/11 [==============================] - 0s 16ms/step - loss: 538.9026 - mse: 538.9026 - val_loss: 601.6938 - val_mse: 601.6938
Epoch 5/10
11/11 [==============================] - 0s 18ms/step - loss: 528.5445 - mse: 528.5445 - val_loss: 589.8356 - val_mse: 589.8356
Epoch 6/10
11/11 [==============================] - 0s 16ms/step - loss: 518.9130 - mse: 518.9130 - val_loss: 576.8784 - val_mse: 576.8784
Epoch 7/10
11/11 [==============================] - 0s 27ms/step - loss: 506.5880 - mse: 506.5880 - val_loss: 563.4048 - val_mse: 563.4048
Epoch 8/10
11/11 [=========

In [9]:
# model evaluation
model.evaluate(X_test_scaled, y_test)

4/4 [==============================] - 0s 4ms/step - loss: 486.4476 - mse: 486.4476


[486.4476013183594, 486.4476013183594]

This model has a MSE of around $486.4476$.

# HyperParameters
The Hyperparameters class is used to specify a set of hyperparameters and their values, to be used in the model building function.

* **HyperParameters class**
* **Boolean method**
* **Choice method**
* **Fixed method**
* **Float method**
* **Int method**
* **conditional_scope method**
* **get method**

## HyperParameter Class

In [10]:
keras_tuner.HyperParameters()

**Attributes**
* **values**: A dict mapping hyperparameter names to current values.

## Boolean Method
Choice between True and False

**Arguments**
* **name**: A string. the name of parameter. Must be unique for each HyperParameter instance in the search space.
* **default**: Boolean, the default value to return for the parameter. If unspecified, the default value will be False.
* **parent_name**: Optional string, specifying the name of the parent HyperParameter to use as the condition to activate the current HyperParameter.
* **parent_values**: Optional list of the values of the parent HyperParameter to use as the condition to activate the current HyperParameter.

**Returns**

The value of the hyperparameter, or None if the hyperparameter is not active.

## Choice Method
Choice of one value among a predefined set of possible values.

**Arguments**
* **name**: A string. the name of parameter. Must be unique for each HyperParameter instance in the search space.
* **values**: A list of possible values. Values must be int, float, str, or bool. All values must be of the same type.
* **ordered**: Optional boolean, whether the values passed should be considered to have an ordering. Defaults to True for float/int values. Must be False for any other values.
* **default**: Optional default value to return for the parameter. If unspecified, the default value will be: - None if None is one of the choices in values - The first entry in values otherwise.
* **parent_name**: Optional string, specifying the name of the parent HyperParameter to use as the condition to activate the current HyperParameter.
* **parent_values**: Optional list of the values of the parent HyperParameter to use as the condition to activate the current HyperParameter.

**Returns**

The value of the hyperparameter, or None if the hyperparameter is not active.

## Fixed Method
Fixed, untunable value.

**Arguments**
* **name**: A string. the name of parameter. Must be unique for each *HyperParameter* instance in the search space.
* **value**: The value to use (can be any JSON-serializable Python type).
* **parent_name**: Optional string, specifying the name of the parent *HyperParameter* to use as the condition to activate the current *HyperParameter*.
* **parent_values**: Optional list of the values of the parent HyperParameter to use as the condition to activate the current HyperParameter.

**Returns**

The value of the hyperparameter, or None if the hyperparameter is not active.

## Float Method
Floating point range, can be evenly divided.

**Arguments**
* **name**: A string. the name of parameter. Must be unique for each HyperParameter instance in the search space.
* **min_value**: Float, the lower bound of the range.
* **max_value**: Float, the upper bound of the range.
* **step**: Optional float, e.g. 0.1, the smallest meaningful distance between two values. Whether step should be specified is Oracle dependent, since some Oracles can infer an optimal step automatically.
* **sampling**: Optional string. One of "linear", "log", "reverse_log". Acts as a hint for an initial prior probability distribution for how this value should be sampled, e.g. "log" will assign equal probabilities to each order of magnitude range.
* **default**: Float, the default value to return for the parameter. If unspecified, the default value will be min_value.
* **parent_name**: Optional string, specifying the name of the parent HyperParameter to use as the condition to activate the current HyperParameter.
* **parent_values**: Optional list of the values of the parent HyperParameter to use as the condition to activate the current HyperParameter.

**Results**

The value of the hyperparameter, or None if the hyperparameter is not active.

## Int Method
Integer range.

Note that unlike Python's range function, max_value is included in the possible values this parameter can take on.

**Arguments**

* **name**: A string. the name of parameter. Must be unique for each HyperParameter instance in the search space.
* **min_value**: Integer, the lower limit of range, inclusive.
* **max_value**: Integer, the upper limit of range, inclusive.
* **step**: Integer, the distance between two consecutive samples in the range. Defaults to 1.
* **sampling**: Optional string. One of "linear", "log", "reverse_log". Acts as a hint for an initial prior probability distribution for how this value should be sampled, e.g. "log" will assign equal probabilities to each order of magnitude range.
* **default**: Integer, default value to return for the parameter. If unspecified, the default value will be min_value.
* **parent_name**: Optional string, specifying the name of the parent HyperParameter to use as the condition to activate the current HyperParameter.
* **parent_values**: Optional list of the values of the parent HyperParameter to use as the condition to activate the current HyperParameter.

**Returns**

The value of the hyperparameter, or None if the hyperparameter is not active.

## Conditional_scope Method
Opens a scope to create conditional HyperParameters.

All **HyperParameters** created under this scope will only be active when the parent **HyperParameter** specified by **parent_name** is equal to one of the values passed in **parent_values**.

When the condition is not met, creating a **HyperParameter** under this scope will register the **HyperParameter**, but will return **None** rather than a concrete value.

Note that any Python code under this scope will execute regardless of whether the condition is met.

This feature is for the **Tuner** to collect more information of the search space and the current trial. It is especially useful for model selection. If the parent **HyperParameter** is for model selection, the **HyperParameters** in a model should only be active when the model selected, which can be implemented using **conditional_scope**.

**Arguments**

* **parent_name**: A string, specifying the name of the parent HyperParameter to use as the condition to activate the current HyperParameter.
* **parent_values**: A list of the values of the parent HyperParameter to use as the condition to activate the current HyperParameter.


# Tuning with Keras Tuner
To start tuning the model in keras tuner, let’s define a hypermodel first. **Hypermodel** is a keras tuner class that lets you define the model with a searchable space and build it.

In [23]:
from keras_tuner import HyperModel

In [24]:
class RegressionHyperModel(HyperModel):
    def __init__(self, input_shape):
        self.input_shape = input_shape
    def build(self, hp):
        model = keras.Sequential()
        model.add(layers.Dense(units=hp.Int('units',min_value = 8, max_value = 64, step = 4, default=8),
                activation=hp.Choice('dense_activation',values=['relu', 'tanh', 'sigmoid'],default='relu'),
                input_shape=input_shape))
        
        model.add(layers.Dense(units=hp.Int('units', min_value = 16, max_value = 64, step = 4, default=16),
                activation=hp.Choice('dense_activation',values=['relu', 'tanh', 'sigmoid'],default='relu')))
        
        model.add(layers.Dropout(hp.Float('dropout',min_value=0.0,max_value=0.1,default=0.005,step=0.01)))
        
        model.add(layers.Dense(1))
        
        model.compile(optimizer='rmsprop',loss='mse',metrics=['mse'])
        
        return model

You may have noticed **hp.Int**, **hp.Float**, and **hp.Choice**, these are used to define a search space for a hyperparameter that accepts an integer, float and a category respectively. A complete list of hyperparameter methods can be found above. **'hp'** is an alias for Keras Tuner's HyperParameters class.

Hyperparameter such as the number of units in a dense layer accepts an integer, hence, **hp.Int** is used to define a range of integers to try. Similarly, the dropout rate accepts a float value so **hp.Float** is used. Both **hp.Int** and **hp.Float** requires a name, minimum value and maximum value, while the step size and default value is optional.

The **hp.Int** search space below is named, **"units"**, and will have values from $8$ to $64$ in multiples of $4$, and a default value of $8$. **hp.Float** is used similarly as **hp.Int** but accepts float values.

**hp.Choice** is used to define a categorical hyperparameter such as the activation function. The search space below, named "dense_activation", will choose between "relu", "tanh", and "sigmoid" functions, with a default value set to "relu".

## Instantiate HyperModel
Let’s instantiate a hypermodel object. Input shape varies per dataset and the problem you are trying to solve.

In [25]:
input_shape = (X_train.shape[1],)
hypermodel = RegressionHyperModel(input_shape)

### Random Search
As the name suggests, this hyperparameter tuning method randomly tries a combination of hyperparameters from a given search space. To use this method in keras tuner, let’s define a tuner using one of the available Tuners.

In [26]:
tuner_rs = keras_tuner.RandomSearch(hypermodel, objective='mse', seed=42, max_trials=10, executions_per_trial=2)

INFO:tensorflow:Reloading Tuner from .\untitled_project\tuner0.json


Run the random search tuner using the search method.

In [27]:
tuner_rs.search(X_train_scaled, y_train, epochs=10, validation_split=0.2, verbose=1)

INFO:tensorflow:Oracle triggered exit


In the next cells, we have retrieved the best model and used it to evaluate performance on the test dataset which we had used as a validation dataset. Then, we have printed the tuning summary as well.

In [28]:
best_params = tuner_rs.get_best_hyperparameters()

best_params[0].values

{'units': 60, 'dense_activation': 'relu', 'dropout': 0.1}

In [29]:
best_model = tuner_rs.get_best_models(num_models=1)[0]
loss, mse = best_model.evaluate(X_test_scaled, y_test)

4/4 [==============================] - 1s 4ms/step - loss: 50.6940 - mse: 50.6940


Random search's MSE is $50.69$, a very big improvement from not performing any tuning at all.

### Hyperband
Hyperband is based on the algorithm by Li et. al. It optimizes random search method through adaptive resource allocation and early-stopping. Hyperband first runs random hyperparameter configurations for one iteration or two, then selects which configurations perform well, then continues tuning the best performers.

In [30]:
tuner_hb = keras_tuner.Hyperband(hypermodel, max_epochs=5, objective='mse', seed=42, hyperband_iterations=2, overwrite = True)
tuner_hb.search(X_train_scaled, y_train, epochs=10, validation_split=0.2, verbose=1)

Trial 20 Complete [00h 00m 03s]
mse: 530.1045532226562

Best mse So Far: 215.310302734375
Total elapsed time: 00h 01m 03s
INFO:tensorflow:Oracle triggered exit


In the next cells, we have retrieved the best model and used it to evaluate performance on the test dataset which we had used as a validation dataset. Then, we have printed the tuning summary as well.

In [31]:
best_params = tuner_hb.get_best_hyperparameters()

best_params[0].values

{'units': 60,
 'dense_activation': 'relu',
 'dropout': 0.1,
 'tuner/epochs': 5,
 'tuner/initial_epoch': 2,
 'tuner/bracket': 1,
 'tuner/round': 1,
 'tuner/trial_id': '0004'}

In [32]:
best_model = tuner_hb.get_best_models(num_models=1)[0]
best_model.evaluate(X_test_scaled, y_test)

4/4 [==============================] - 0s 3ms/step - loss: 207.0387 - mse: 207.0387


[207.03868103027344, 207.03868103027344]

The resulting MSE is $207.03$ which is a lot worse when compared to random search but a little bit better than not tuning at all.

### Bayesian Optimization
Bayesian optimization is a probabilistic model that maps the hyperparameters to a probability score on the objective function. Unlike Random Search and Hyperband models, Bayesian Optimization keeps track of its past evaluation results and uses it to build the probability model.

In [33]:
tuner_bo = keras_tuner.BayesianOptimization(hypermodel, objective='mse', max_trials=10, seed=42, executions_per_trial=2, overwrite = True)
tuner_bo.search(X_train_scaled, y_train, epochs=10, validation_split=0.2, verbose=1)

Trial 10 Complete [00h 00m 08s]
mse: 42.36796474456787

Best mse So Far: 42.36796474456787
Total elapsed time: 00h 01m 27s
INFO:tensorflow:Oracle triggered exit


In the next cells, we have retrieved the best model and used it to evaluate performance on the test dataset which we had used as a validation dataset. Then, we have printed the tuning summary as well.

In [34]:
best_params = tuner_bo.get_best_hyperparameters()

best_params[0].values

{'units': 64, 'dense_activation': 'relu', 'dropout': 0.05}

In [35]:
best_model = tuner_bo.get_best_models(num_models=1)[0]
best_model.evaluate(X_test_scaled, y_test)

4/4 [==============================] - 0s 4ms/step - loss: 32.3583 - mse: 32.3583


[32.35834884643555, 32.35834884643555]

Best model MSE tuned using Bayesian optimization is $32.358$, better than the first two tuners we have tried.